src/buildstream/source.py

```diff
@@ -684,6 +821,19 @@ def is_cached(self) -> bool:
         """
         raise ImplError("Source plugin '{}' does not implement is_cached()".format(self.get_kind()))
 
+    def collect_source_info(self) -> Iterable[SourceInfo]:
+        """Get the :class:`.SourceInfo` objects describing this source
+
+        This method is guaranteed to only be called whenever
+        :func:`Source.is_resolved() <buildstream.source.Source.is_resolved>`
+        returns `True`.
+
+        Returns: the :class:`.SourceInfo` objects describing this source
+
+        *Since: 2.5*
+        """
+        raise ImplError("Source plugin '{}' does not implement collect_source_info()".format(self.get_kind()))
```

Comment:
Could we have this collect source info from source fetchers by default? Any plugin that uses source fetchers will definitely want to implement source info in source fetchers.

Of course, this is more for standardisation than anything, one can always implement it like you did in https://github.com/apache/buildstream-plugins/pull/87

Reply:
This seems sensible, I'll add something to the `SourceFetcher` API for that.

Reply:
I've added this and updated the plugins which use fetchers.

In addition, I've added a public facing `SourceImplError` and updated the `collect_source_info()` docstring to specify that callers of the API need to handle `SourceImplError` exceptions.